pkg/controller/jobframework/reconciler.go

```diff
@@ -460,15 +460,17 @@ func (r *JobReconciler) ReconcileGenericJob(ctx context.Context, req ctrl.Reques
 	// handle a job when waitForPodsReady is enabled, and it is the main job
 	if r.waitForPodsReady {
 		log.V(3).Info("Handling a job when waitForPodsReady is enabled")
-		condition := generatePodsReadyCondition(job, wl)
-		// optimization to avoid sending the update request if the status didn't change
-		if !apimeta.IsStatusConditionPresentAndEqual(wl.Status.Conditions, condition.Type, condition.Status) {
-			log.V(3).Info(fmt.Sprintf("Updating the PodsReady condition with status: %v", condition.Status))
+		condition := generatePodsReadyCondition(log, job, wl)
+		if !workload.HasCondition(wl, &condition) {
```

Comment:
I think the semantics of the HasCondition is quite unclear just by looking at its name. In particular, should we check timestamps, message, reason etc. are the obvious questions. 

I think it would be better to have HasConditionWithTypeAndReason (I believe we should already have something like this). 

I did a quick check and the HasCondition function checks the message, I think we should avoid building any logic on the messages. They are only for humans.

Reply:
> There are two cases in which the message is the only differentiator between conditions:

Thanks for explaining, I see the motivation.

> Are you suggesting merging those two and setting more generic message?

Possibly, but not necessarily. I can see at least two options:
1. make the message generic IIUC this is how it was done before 
2. discriminate against the two states by the reason field, e.g. `WaitForPodsStart` and `NotAdmitted`

Basically, we should not build any logic based on messages being equal or different, because messages are free to change at any point. We have the analogous policy in the code k8s.

My preference is for (2.), but since it was done like in (1.) I'm ok to leave it for follow up.

Reply:
I also like the second approach better. I've added `NotAdmitted` reason and removed message comparison from the `HasConditionWithTypeAndReason` function